### ISSUE TITLE
cron halts part way through when history arrays are large

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#10386]
This isn't a new bug and it's been mentioned in other issues (e.g., https://github.com/HabitRPG/habitica/issues/7727, https://github.com/HabitRPG/habitica/issues/7909, https://github.com/HabitRPG/habitica/issues/9068) but it's still happening so I think it's worth having a new issue opened for it.

When a player has Habits or Dailies with large `history` arrays or if their `user.history` array is large, cron will sometimes halt part way through, presumably because it crashes when adding to or preening the arrays. Sometimes Loggly shows VersionErrors but not always.

Admins can fix this for individual players when it's reported by emptying the arrays. Cron then runs normally on the next day. My personal feeling for when the arrays start to cause problems is around 300 entries.

We could fix this automatically by having cron purge the history arrays when they're too large, but that's not ideal, especially for subscribers who have paid to have history kept longer.

Another option would be to wrap try-catch blocks or similar error handling around the cron code that processes individual tasks so that if one task causes a failure, the rest of cron can continue (including processing the other tasks). That would also make cron more robust for any other current or future problems that might cause failures. I don't know how possible this would be but I think it might be worth investigating, rather than purging the history prematurely.

I believe the staff have plans to keep improving Habitica's architecture to prevent VersionErrors and similar problems, but I don't know the timeframes on that so a workaround now might be helpful.

I'm marking this suggestion-discussion for us to determine the best approach.

**Comment by SabreCat** [comment]
Very high volume issue in bug reporting channels, so I'm marking this Important despite the "discussion needed" status.

**Comment by Alys** [comment]
Reopening since this will still happen from large history arrays.

**Comment by paglias** [comment]
Maybe we should wait for new reports after the migration before reopening
this?

Il Sab 23 Giu 2018, 08:30 Alys <notifications@github.com> ha scritto:

**Comment by Alys** [comment]
If you want sure, I'll close it now, but some of the original reports were from the Dailies and user.history arrays.

**Comment by Alys** [comment]
I think we've suspected for a while that this might still be happening (I've found some large history entries in people reporting cron failures and I _think_ SabreCat has too but I might be misremembering). I'm more certain now so I'm reopening this. @NookaBean (77468ba2-c08a-4c12-9cb8-70c6a7ee437c) has reported:

"There was some kind of error message while I started my new day. Now it's saying I didn't find any moonstones yesterday. Last time I (double) checked a few minutes ago, I'd reached my target of finding eleven, so I should have completed the quest."

[Loggly shows this](https://habitica.loggly.com/search#terms=77468ba2-c08a-4c12-9cb8-70c6a7ee437c&from=2019-02-01T00:00:00.000Z&until=2019-02-17T04:21:32.440Z&source_group=&event=062b5d10-3247-11e9-8080-06e403ed47d2,1550361623501,137696544239082768):
```
...
httpCode:500
isHandledError:false
level:error
message: **VersionError: No matching document found for id "0f89f02b-36ac-401d-a5f9-b13dfd595ae3"** version 818 modifiedPaths "value, streak, history, nextDue"
    at VersionError.MongooseError [as constructor] (/usr/src/habitrpg/node_modules/mongoose/lib/error/mongooseError.js:13:11)
    at new VersionError (/usr/src/habitrpg/node_modules/mongoose/lib/error/version.js:18:17)
    at generateVersionError (/usr/src/habitrpg/node_modules/mongoose/lib/model.js:398:10)
    at model.Model.save (/usr/src/habitrpg/node_modules/mongoose/lib/model.js:461:27)
    at tasks.forEach.task (/usr/src/habitrpg/website/transpiled-babel/middlewares/cron.js:119:47)
    at Array.forEach (<anonymous>)
    at cronAsync (/usr/src/habitrpg/website/transpiled-babel/middlewares/cron.js:118:11)
    ...
```

That ID number points to a Daily that had 362 `history` entries, which is certainly in the range that we found was causing this bug in the past. 

I've deleted the history entries for that task and others for this player to prevent the problem happening again for now.

**Comment by paglias** [comment]
I'm not sure that's the case because VersionError(s) are not a generic type
of error but they are only thrown manually by mongoose when you try to
update a document in parallel. It's more likely having clicked the start my
day button without the request for scoring a daily being complete

Il giorno Dom 17 Feb 2019, 05:34 Alys <notifications@github.com> ha scritto:

**Comment by Alys** [comment]
Cool, so preventing the start my day button being clicked while a sync is in progress would stop that kind of error happening. Is that something that can be done?

**Comment by paglias** [comment]
Yeah it can be done but it'll have to be a platform specific change (so a
fix each for web, iOS and Android)

Il giorno Dom 17 Feb 2019, 13:58 Alys <notifications@github.com> ha scritto:
